packages/@aws-cdk/core/lib/private/synthesis.ts

```diff
@@ -110,6 +111,41 @@ function prepareTree(root: IConstruct) {
   visit(root, 'post', construct => construct.onPrepare());
 }
 
+/**
+ * Find all stacks and add Metadata Resources to all of them
+ *
+ * There is no good generic place to do this. Can't do it in the constructor
+ * (because adding a child construct makes it impossible to set context on the
+ * node), and the generic prepare phase is deprecated.
+ *
+ * Only do this on [parent] stacks (not nested stacks), don't do this when
+ * disabled by the user.
+ *
+ * Also, only when running via the CLI. If we do it unconditionally,
+ * all unit tests everywhere are going to break massively. I've spent a day
+ * fixing our own, but downstream users would be affected just as badly.
+ *
+ * Stop at Assembly boundaries.
+ */
+function injectMetadataResources(root: IConstruct) {
+  visit(root, 'post', construct => {
+    // Only on top-level stacks
+    if (!Stack.isStack(construct)
+      || construct.parentStack
+      // Unless disabled
+      || construct.node.tryGetContext(cxapi.DISABLE_VERSION_REPORTING)
+      // While running via CLI
+      || !process.env[cxapi.CLI_VERSION_ENV]) { return; }
```

Comment:
Why?

Reply:
I believe I described that twice.

Reply:
Won't it make sense to just not include the CLI version in case this environment variable in not defined?

Reply:
It's not actually about the `CLI_VERSION_ENV` itself. I'm using `CLI_VERSION_ENV` as a trigger to determine whether we're running under the CLI (and hence emitting the metadata)